store/v2/commitment/iavlv2/tree.go

```diff
@@ -92,57 +96,97 @@ func (t *Tree) Get(version uint64, key []byte) ([]byte, error) {
 	if err := isHighBitSet(version); err != nil {
 		return nil, err
 	}
-	if int64(version) != t.tree.Version() {
-		cloned, err := t.tree.ReadonlyClone()
-		if err != nil {
-			return nil, err
-		}
-		if err = cloned.LoadVersion(int64(version)); err != nil {
-			return nil, err
-		}
-		return cloned.Get(key)
-	} else {
-		return t.tree.Get(key)
+	v := int64(version)
+	h := t.tree.Version()
+	if v > h {
+		return nil, fmt.Errorf("get: cannot read future version %d; h: %d path=%s", v, h, t.path)
+	}
+	versionFound, val, err := t.tree.GetRecent(v, key)
+	if versionFound {
+		return val, err
 	}
+	cloned, err := t.tree.ReadonlyClone()
+	if err != nil {
+		return nil, err
+	}
+	if err = cloned.LoadVersion(int64(version)); err != nil {
+		return nil, err
+	}
+	return cloned.Get(key)
+}
+
+func (t *Tree) Has(version uint64, key []byte) (bool, error) {
+	res, err := t.Get(version, key)
+	return res != nil, err
 }
 
 func (t *Tree) Iterator(version uint64, start, end []byte, ascending bool) (corestore.Iterator, error) {
 	if err := isHighBitSet(version); err != nil {
 		return nil, err
 	}
-	if int64(version) != t.tree.Version() {
-		return nil, fmt.Errorf("loading past version not yet supported")
+	h := t.tree.Version()
+	v := int64(version)
+	if v > h {
+		return nil, fmt.Errorf("iterator: cannot read future version %d; h: %d", v, h)
+	}
+	ok, itr := t.tree.IterateRecent(v, start, end, ascending)
+	if ok {
+		return itr, nil
+	}
+	cloned, err := t.tree.ReadonlyClone()
+	if err != nil {
+		return nil, err
+	}
+	if err = cloned.LoadVersion(int64(version)); err != nil {
+		return nil, err
 	}
 	if ascending {
+		// inclusive = false is IAVL v1's default behavior.
+		// the read expectations of certain modules (like x/staking) will cause a panic if this is changed.
 		return t.tree.Iterator(start, end, false)
 	} else {
 		return t.tree.ReverseIterator(start, end)
 	}
 }
 
 func (t *Tree) Export(version uint64) (commitment.Exporter, error) {
-	return nil, errors.New("snapshot import/export not yet supported")
+	if err := isHighBitSet(version); err != nil {
+		return nil, err
+	}
+	e, err := t.tree.Export(int64(version), iavl.PostOrder)
+	if err != nil {
+		return nil, err
+	}
+	return &Exporter{e}, nil
 }
 
 func (t *Tree) Import(version uint64) (commitment.Importer, error) {
-	return nil, errors.New("snapshot import/export not yet supported")
+	if err := isHighBitSet(version); err != nil {
+		return nil, err
+	}
+	importer, err := t.tree.Import(int64(version))
+	if err != nil {
+		return nil, err
+	}
+	return &Importer{importer}, nil
 }
 
 func (t *Tree) Close() error {
 	return t.tree.Close()
 }
 
 func (t *Tree) Prune(version uint64) error {
-	if err := isHighBitSet(version); err != nil {
-		return err
-	}
-
-	return t.tree.DeleteVersionsTo(int64(version))
+	// do nothing, IAVL v2 has its own advanced pruning mechanism
+	return nil
```

Comment:
what is the design here for configuring pruning setting? how does a node operator say they want to prune up to x?

Reply:
The `keep-recent` pruning option is passed to iavl v2.

https://github.com/cosmos/cosmos-sdk/blob/200348b598d17c83ed10dcbeecdfba821ee39ed3/simapp/v2/app.go#L230